File: python/0028实现strStr.py


# 暴力
def strStr(haystack:str,needle:str):
    
    if(not needle):
        return 0
    h_len=len(haystack)
    n_len=len(needle)
    if(h_len<n_len):
        return -1

    for i,s in enumerate(haystack):
        if(s==needle[0]):
            for j,k in enumerate(needle):
                if((i+j)>=h_len):
                    return -1
                if(haystack[i+j]!=k):
                    break
                if(j==n_len-1):
                    return i
    return -1
    

print(strStr("mississippi","issipi"))

File: cpp/0028实现strStr.cpp
#include <vector>
#include<iostream>

using namespace std;

int strStr(string haystack, string needle)
{
    int h_size = haystack.size();
    int n_size = needle.size();
    if(n_size==0)
        return 0;
    if(h_size<n_size)
        return -1;

    for (int i = 0;i<h_size;++i)
    {
        int j = 0;
        
        if(haystack[i]==needle[0])
        {
            j = 1;
            
            for (; j < n_size; ++j)
            {
                if (i + j >= h_size)
                    return -1;
             
                if (haystack[i + j] != needle[j])
                    break;
            }

        if (j == n_size)
            return i;
        }
    }

    return -1;
}

int main()
{
    string h = "asd";
    string n = "sd";

    cout<<strStr(h,n);
}